### DISCUSSION TITLE
HTML output and pdf output?

### DISCUSSION THREAD

**Discussion by ericvmai** [quarto-dev/quarto-cli#2397] [Q&A]
I am trying to render my Quarto project to both html and pdf.

My colleague has shown that a leaflet map will render to pdf without issue here (https://github.com/knopfjt/sandbox/blob/main/quarto-test/docs/Quarto-Test.pdf) although the leaflet map's image is messed up.

But when I try to render my Quarto project located here (https://github.com/hdadvisors/chesterfield-market-analysis/tree/quarto-transition), I get the following error:

> Error: Functions that produce HTML output found in document targeting pdf output.
> document targeting pdf output. 
> Please change the output type of this document to HTML. Alternatively, you can allow
> HTML output in non-HTML formats by adding this option to the YAML front-matter of
> your rmarkdown file:
> 
>   always_allow_html: true
> 
> Note however that the HTML output will not be visible in non-HTML formats.

Anyone have any solutions here?

**Comment by rgaiacs** [comment]
Hi @ericvmai,

Some R and Python objects has different representation depending of the final format that you are producing. Unfortunately, even the objects that has different representation might produce the "wrong" output given a couple of different reasons.

In my documents, I'm using the follow approach for images:


````
::: {#fig-awesome}
::: {.content-visible unless-format="html"}
```{r}
#| output: true

data_vis |> to_png()
```
:::

::: {.content-visible when-format="html"}
```{r}
#| output: true

data_vis |> to_html()
```
:::

Awesome figure.
:::
````

This allows me to hand craft the correct settings for a PNG/PDF to be used when creating the PDF and for the HTML to be used when creating the HTML.

I need to have

```
---
always_allow_html: true
---
```

at the document YAML header!

**Reply by ericvmai** [reply]
Thanks. This makes sense to me. But unsure of how to apply it to my quarto project in R! Are the :::{} you use a function of quarto?

**Reply by rgaiacs** [reply]
The `:::`is a Pandoc markup for HTML `<div>`. Read more at https://quarto.org/docs/authoring/conditional.html.

**Comment by cderv** [comment]
This type of issue should only happens if you have HTML only content in your document. In that case, they should either be ignored, or included somehow. 

Leaflet Maps or other HTML widgets should be handled directly by **knitr** by taking a screenshort of the map or other HTML visualisation. So `always_allow_html` is not needed. 

@ericvmai Can you think of any HTML only output you may have in your project ? 

Conditional solution suggested by @rgaiacs is an option for such content if the tool used does not handle it (example of **kableExtra** used with Bootstrap styling - this would output HTML and could not be included in PDF, where LaTeX tables are needed)

So overall, which files is throwing the issue, and what type of content does this file have ? That will help you know how to continue your conversion
